### PR TITLE
reader-writer differentiation in mtev_conf

### PR DESCRIPTION
**Pull request by postwait**

### CODE REVIEW

File: src/mtev_conf.c [hunk 1]
```diff
@@ -57,6 +64,10 @@
 #include "mtev_str.h"
 #include "mtev_thread.h"
 
+#if (__STDC_VERSION__ >= 201112L)
```
Comment (cody271) [inline src/mtev_conf.c]: What does this do here?

File: src/mtev_conf.c [hunk 2]
```diff
@@ -79,18 +90,111 @@ const char *_mtev_version = MTEV_VERSION;
...
 
-static pthread_mutex_t global_config_lock;
+/* This is tragically ugly.
+ *
+ * Ideally we would use a pthread_rwlock, however those are aware of only
+ * pthreads and we have aco threads to worry about as well.  We also need
+ * recursive reader locks so we use ck_rwlock_recursive_t locks here but
+ * they don't support recursive readers (despite having a call for that).
+ *
+ * The "work" but if you have a read lock and attempt to recursively acquire
```
Comment (cody271) [inline src/mtev_conf.c]:
```suggestion
 * They "work" but if you have a read lock and attempt to recursively acquire
```

File: src/mtev_conf.c [hunk 3]
```diff
@@ -79,18 +90,111 @@ const char *_mtev_version = MTEV_VERSION;
...
+ *
+ * In order to make the readlocks recursive we need a ACO-aware thread-local
+ * recursion counter for this lock and the whole global_recursion_{inc,dec}
+ * and _g(un)lock_read routines manage this.
+ *
+ * _glock is simpler b/c recursive write locks work fine.
+ *
+ * There still exist possible deadlock if the callers are not careful.
+ *
+ * If you old a read lock in an ACO thread and do something that yields
```
Comment (cody271) [inline src/mtev_conf.c]:
```suggestion
 * If you hold a read lock in an ACO thread and do something that yields
```

File: src/mtev_thread.c [hunk 4]
```diff
@@ -95,6 +96,12 @@ static mtev_boolean mtev_disable_binding = mtev_false;
 uint32_t
 mtev_thread_id(void) {
   if(!mtev_current_thread_id) mtev_current_thread_id = getthreadid();
+  aco_t *co = aco_co();
```
Comment (cody271) [inline src/mtev_thread.c]: Should this use the new `aco_co_thread()` ?

Comment (postwait) [reply src/mtev_thread.c]: yup

File: src/mtev_conf.c (hunk 1 above)
Comment (postwait) [reply src/mtev_conf.c]: _Static_assert isn't defined before c11.

**Pull request merged**